llvm/lib/Support/SHA1.cpp

```diff
@@ -307,3 +307,27 @@ std::array<uint8_t, 20> SHA1::hash(ArrayRef<uint8_t> Data) {
   memcpy(Arr.data(), S.data(), S.size());
   return Arr;
 }
+
+std::string SHA1::hashToString(ArrayRef<uint8_t> Data, StringRef Mapping) {
+  constexpr auto BitsPerChar = 6;
+  constexpr auto CombinationPerChar = 1 << BitsPerChar;
+  constexpr auto LoopIterations = 27;
+  assert(Mapping.size() >= CombinationPerChar);
+  std::array<uint8_t, 20> Hash = SHA1::hash(Data);
+  std::string Res;
+  Res.reserve(LoopIterations);
+
+  uint16_t Value = Hash[1] << 8 | Hash[0];
+  for (unsigned Bits = 0, Elem = 1; Res.size() < LoopIterations;) {
+    Res.push_back(Mapping[Value & 0b111111]);
+    Bits += BitsPerChar;
+    Value >>= BitsPerChar;
+    if (Bits >= 8) {
+      Elem++;
+      Bits -= 8;
+      if (LLVM_LIKELY(Elem < Hash.size()))
+        Value |= Hash[Elem] << (8 - Bits);
+    }
```

Comment:
I found this still difficult to understand because `Bits` which is not physical. What about:
```suggestion
  uint16_t Value = Hash[0];
  for (auto AvailableBits = CHAR_BITS, Elem = 1; AvailableBits > 0;) {
    Res.push_back(Mapping[Value & ((1 << BitsPerChar) - 1)]);
    AvailableBits -= BitsPerChar;
    Value >>= BitsPerChar;
    if (AvailableBits < BitsPerChar && Elem < Hash.size()) {
        Value |= Hash[Elem++] << AvailableBits;
        AvailableBits += CHAR_BITS;
    }
```
I think now we have reached the bottom line here.
This is a rehearsal for your future coding interviews... :-)